File: src/main/java/top/chiloven/lukosbot2/lifecycle/ConsoleLifecycle.java
package top.chiloven.lukosbot2.lifecycle;

import lombok.extern.log4j.Log4j2;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.SmartLifecycle;
import org.springframework.stereotype.Service;
import top.chiloven.lukosbot2.core.cli.CliCmdContext;
import top.chiloven.lukosbot2.core.cli.CliCmdProcessor;
import top.chiloven.lukosbot2.util.concurrent.Execs;

import java.io.*;
import java.util.concurrent.ExecutorService;

@Service
@ConditionalOnProperty(prefix = "lukos.cli", name = "enabled", havingValue = "true")
@Log4j2
public class ConsoleLifecycle implements SmartLifecycle {

    private final CliCmdContext context;
    private final CliCmdProcessor processor;

    private volatile boolean running;
    private ExecutorService executor;

    public ConsoleLifecycle(CliCmdProcessor processor) {
        this.processor = processor;
        this.context = new CliCmdContext(System.out);
    }

    @Override
    public void start() {
        if (running) return;
        running = true;

        executor = Execs.newVirtualExecutor("cli-");
        executor.submit(() -> {
            try (var reader = new BufferedReader(new InputStreamReader(new NonClosingInputStream(System.in)))) {
                while (running && !Thread.currentThread().isInterrupted()) {
                    String line = reader.readLine();
                    if (line == null) break;
                    if (line.isBlank()) continue;

                    try {
                        processor.handle(line, context);
                    } catch (Exception e) {
                        log.warn("[Cli] Failed to execute command: {}", e.getMessage(), e);
                    }
                }
            } catch (Exception e) {
                if (running && !isExpectedShutdown(e)) {
                    log.warn("[Cli] Console lifecycle stopped unexpectedly: {}", e.getMessage(), e);
                }
            } finally {
                running = false;
            }
        });
    }

    @Override
    public void stop() {
        running = false;
        if (executor != null) {
            executor.shutdownNow();
            executor = null;
        }
    }

    @Override
    public boolean isRunning() {
        return running;
    }

    private static boolean isExpectedShutdown(Exception e) {
        if (e instanceof IOException io && "Stream closed".equalsIgnoreCase(io.getMessage())) {
            return true;
        }
        return e instanceof InterruptedException;
    }

    @Override
    public void stop(Runnable callback) {
        stop();
        callback.run();
    }

    @Override
    public int getPhase() {
        return 0;
    }

    private static final class NonClosingInputStream extends FilterInputStream {

        private NonClosingInputStream(InputStream in) {
            super(in);
        }

        @Override
        public void close() {
            // Never close System.in, otherwise a whole-process restart cannot recreate the console lifecycle.
        }

    }

}

File: app/src/main/java/top/chiloven/lukosbot2/lifecycle/ConsoleLifecycle.kt
/*
 * Copyright © 2026 Chiloven945
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package top.chiloven.lukosbot2.lifecycle

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.isActive
import kotlinx.coroutines.runInterruptible
import org.apache.logging.log4j.LogManager
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.SmartLifecycle
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.core.BotCoroutineRuntime
import top.chiloven.lukosbot2.core.command.cli.CliCmdContext
import top.chiloven.lukosbot2.core.command.cli.CliCmdProcessor
import java.io.*

@Service
@ConditionalOnProperty(
    prefix = "lukos.cli", name = ["enabled"], havingValue = "true"
)
class ConsoleLifecycle(
    private val processor: CliCmdProcessor,
    private val runtime: BotCoroutineRuntime,
) : SmartLifecycle {

    private val log = LogManager.getLogger(ConsoleLifecycle::class.java)
    private val context = CliCmdContext(System.out)

    @Volatile
    private var running = false
    private var job: Job? = null

    override fun start() {
        if (running) return
        running = true

        job = runtime.launch("console") {
            try {
                val reader = BufferedReader(InputStreamReader(NonClosingInputStream(System.`in`)))
                while (running && isActive) {
                    val line =
                        runInterruptible { reader.readLine() }
                            ?: break
                    if (line.isBlank()) continue

                    try {
                        processor.handle(line, context)
                    } catch (e: Exception) {
                        log.warn(
                            "[Cli] Failed to execute command: {}", e.message, e
                        )
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                if (running && !isExpectedShutdown(e)) {
                    log.warn(
                        "[Cli] Console lifecycle stopped unexpectedly: {}",
                        e.message,
                        e
                    )
                }
            } finally {
                running = false
            }
        }
    }

    override fun stop() {
        running = false
        job?.cancel()
        job = null
    }

    override fun isRunning(): Boolean = running

    override fun stop(callback: Runnable) {
        stop()
        callback.run()
    }

    override fun getPhase(): Int = 0

    private companion object {

        fun isExpectedShutdown(e: Exception): Boolean {
            return e is IOException && "Stream closed".equals(
                e.message, ignoreCase = true
            ) || e is InterruptedException
        }
    }

    /**
     * Never closes System.in, otherwise a whole-process restart cannot recreate the console lifecycle.
     */
    private class NonClosingInputStream(inner: InputStream) : FilterInputStream(inner) {

        override fun close() {
            // no-op
        }

    }

}
